### PR TITLE
[RISCV] Add ability to list extensions enabled for a target

### PR DESCRIPTION
**Pull request by michaelmaitland**
bb83a3d introduced `--print-enabled-extensions` command line option for AArch64. This patch introduces RISC-V support for this option.

`riscvExtensionsHelp` is renamed to `printSupportedExtensions` to by synonymous with AArch64 and so it is clear what that function does.

### CODE REVIEW

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 1]
```diff
@@ -116,6 +115,45 @@ void llvm::riscvExtensionsHelp(StringMap<StringRef> DescMap) {
...
+void RISCVISAInfo::printEnabledExtensions(
+    bool IsRV64, std::set<StringRef> &EnabledFeatureNames,
+    StringMap<StringRef> &DescMap) {
+  outs() << "Extensions enabled for the given RISC-V target\n\n";
+  PrintExtension("Name", "Version", (DescMap.empty() ? "" : "Description"));
+
+  RISCVISAUtils::OrderedExtensionMap FullExtMap;
+  RISCVISAUtils::OrderedExtensionMap ExtMap;
+  for (const auto &E : SupportedExtensions)
+    if (EnabledFeatureNames.find(E.Name) != EnabledFeatureNames.end()) {
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:127]: You can use `std::set::count`

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 2]
```diff
@@ -116,6 +115,45 @@ void llvm::riscvExtensionsHelp(StringMap<StringRef> DescMap) {
...
+    std::string Version =
+        std::to_string(E.second.Major) + "." + std::to_string(E.second.Minor);
+    PrintExtension(E.first, Version, DescMap[E.first]);
+  }
+
+  outs() << "\nExperimental extensions\n";
+  ExtMap.clear();
+  for (const auto &E : SupportedExperimentalExtensions) {
+    StringRef Name(E.Name);
+    if (EnabledFeatureNames.find("experimental-" + Name.str()) !=
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:141]: `std::set::count`

File: llvm/unittests/TargetParser/RISCVISAInfoTest.cpp [hunk 3]
```diff
@@ -1061,11 +1061,40 @@ For example, clang -march=rv32i_v1p0)";
...
+TEST(TargetParserTest, RISCVPrintEnabledExtensions) {
+  // clang-format off
+  std::string ExpectedOutput =
+R"(Extensions enabled for the given RISC-V target
+
+    Name                 Version   Description
+    i                    2.1       'I' (Base Integer Instruction Set)
+
+Experimental extensions
+    ztso                 0.1       'Ztso' (Memory Model - Total Store Order)
```
Comment (topperc) [inline llvm/unittests/TargetParser/RISCVISAInfoTest.cpp:1082]: ztso is about to become non-experimental

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 4]
```diff
@@ -116,6 +115,44 @@ void llvm::riscvExtensionsHelp(StringMap<StringRef> DescMap) {
...
+    }
+  }
+  for (const auto &E : ExtMap) {
+    std::string Version =
+        std::to_string(E.second.Major) + "." + std::to_string(E.second.Minor);
+    PrintExtension(E.first, Version, DescMap["experimental-" + E.first]);
+  }
+
+  unsigned XLen = IsRV64 ? 64 : 32;
+  outs() << "\nISA String: " << RISCVISAInfo(XLen, FullExtMap).toString();
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:153]: I'm going to remove this RISCVISAInfo constructor in factor of a new static function in RISCVISAInfo. It's only used in lld before this patch and I don't like the division of responsibilities there.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 5]
```diff
@@ -116,6 +115,44 @@ void llvm::riscvExtensionsHelp(StringMap<StringRef> DescMap) {
...
+  for (const auto &E : SupportedExtensions)
+    if (EnabledFeatureNames.count(E.Name) != 0) {
+      FullExtMap[E.Name] = {E.Version.Major, E.Version.Minor};
+      ExtMap[E.Name] = {E.Version.Major, E.Version.Minor};
+    }
+  for (const auto &E : ExtMap) {
+    std::string Version =
+        std::to_string(E.second.Major) + "." + std::to_string(E.second.Minor);
+    PrintExtension(E.first, Version, DescMap[E.first]);
+  }
```
Comment (dtcxzyw) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:135]:
```suggestion
  RISCVISAUtils::OrderedExtensionMap FullExtMap;
  for (const auto &E : SupportedExtensions)
    if (EnabledFeatureNames.count(E.Name) != 0) {
      FullExtMap[E.Name] = {E.Version.Major, E.Version.Minor};
       std::string Version =
        std::to_string(E.Version.Major) + "." + std::to_string(E.Version.Minor);
        PrintExtension(E.Name, Version, DescMap[E.Name]);
    }
```
`SupportedExtensions` is sorted.

Comment (topperc) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:135]: It's sorted but not in the same order as OrderedExtensionMap.

Comment (michaelmaitland) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:135]: I propose that we print with the current order proposed in this patch (canonical). If we decide that we'd like to change printing order, then we should have a follow up PR that changes print order for both `printSupportedExtensions` and `printEnabledExtensions`.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp (hunk 4 above)
Comment (michaelmaitland) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:153]: Updated.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**